Adopt smart pointer in Source/WebKit/UIProcess/API/APIDataTask.cpp

fred-wang opened the pull request:
#### 4ec0648d3c9e9641e57438d2bbe36eb5c4ac4020
<pre>
Adopt smart pointer in Source/WebKit/UIProcess/API/APIDataTask.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=286808">https://bugs.webkit.org/show_bug.cgi?id=286808</a>

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/APIDataTask.cpp:
(API::DataTask::cancel): NetworkProcessProxy::cancelDataTask is non-trivial so hold a smart pointer to NetworkProcessProxy.
WebsiteDataStore::networkProcess()
(API::DataTask::DataTask): WebsiteDataStore::networkProcess() is not trivial, so use protected version of WebPageProxy::websiteDataStore().
sessionID() is not trivial, so use protectedPage().
ProcessThrottler::foregroundActivity()/backgroundActivity() are non trivial, so use protected version of AuxiliaryProcessProxy::throttler().
* Source/WebKit/UIProcess/API/APIDataTask.h:
(API::DataTask::protectedPage const): Introduce protectedPage() method.

Canonical link: <a href="https://commits.webkit.org/289724@main">https://commits.webkit.org/289724@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/950a84f2511b67a66f08296c406b8d42b27bb4c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87796 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42183 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92661 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38546 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7693 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15483 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67789 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/25535 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90798 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5880 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48157 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5667 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37653 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76065 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94548 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14964 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76641 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15219 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75877 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18655 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20245 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18680 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7966 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14980 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20283 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14724 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16506 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/APIDataTask.cpp:
```diff
@@ -47,8 +47,10 @@ void DataTask::setClient(Ref<DataTaskClient>&& client)
 {
-    if (m_networkProcess && m_sessionID && m_identifier)
-        m_networkProcess->cancelDataTask(*m_identifier, *m_sessionID);
+    if (RefPtr networkProcess = m_networkProcess.get()) {
```
I think we should do the opposite to avoid an unnecessary ref/deref if m_session or m_identifier are null opt:
```
if (m_sessionID && m_identifier) {
    if (RefPtr networkProcess = m_networkProcess.get())
        networkProcess->cancelDataTask(*m_identifier, *m_sessionID);
}
```

fred-wang replied:
OK, I was thinking the same but was not sure. I just realized that there are other kind of errors in UIProcess/API/APIDataTask.cpp and that I need to update SaferCPPExpectations, so will resend for review later.

cdumez reviewed (approved):
LGTM with change.

fred-wang commented on Source/WebKit/UIProcess/API/APIDataTask.cpp:
```diff
@@ -62,12 +64,12 @@ DataTask::DataTask(std::optional<WebKit::DataTaskIdentifier> identifier, WeakPtr
     , m_originalURL(WTFMove(originalURL))
-    , m_networkProcess(m_page ? WeakPtr { m_page->websiteDataStore().networkProcess() } : nullptr)
+    , m_networkProcess(m_page ? WeakPtr { m_page->protectedWebsiteDataStore()->networkProcess() } : nullptr)
     , m_sessionID(m_page ? std::optional<PAL::SessionID> { m_page->sessionID() } : std::nullopt)
```
Remaining error is m_page->sessionID() here, but I'm a bit unsure this is "almost trivial" (it just calls m_websiteDataStore->sessionID() with the latter being trivial)?

fred-wang replied:
cc @rniwa

cdumez replied:
It calls `WebPageProxy::sessionID()` which is not inlined and therefore cannot be considered trivial.

cdumez replied:
We either want to make `WebPageProxy::sessionID()` inline or protect the page here. Marking `WebPageProxy::sessionID()` inline may require new header include in WebPageProxy.h, which wouldn't be great.

fred-wang replied:
Thanks I'll protect the page then.

fred-wang committed:
Adopt smart pointer in Source/WebKit/UIProcess/API/APIDataTask.cpp
https://bugs.webkit.org/show_bug.cgi?id=286808

Reviewed by Chris Dumez.

* Source/WebKit/UIProcess/API/APIDataTask.cpp:
(API::DataTask::cancel): NetworkProcessProxy::cancelDataTask is non-trivial so hold a smart pointer to NetworkProcessProxy.
WebsiteDataStore::networkProcess()
(API::DataTask::DataTask): WebsiteDataStore::networkProcess() is not trivial, so use protected version of WebPageProxy::websiteDataStore().
sessionID() is not trivial, so use protectedPage().
ProcessThrottler::foregroundActivity()/backgroundActivity() are non trivial, so use protected version of AuxiliaryProcessProxy::throttler().
* Source/WebKit/UIProcess/API/APIDataTask.h:
(API::DataTask::protectedPage const): Introduce protectedPage() method.

Canonical link: https://commits.webkit.org/289724@main
